### DISCUSSION TITLE
Profiling libresolv.9.dylib symbolization file missing error

### DISCUSSION THREAD

**Discussion by smithht** [idaholab/moose#27735] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I have been using MOOSE for crystal plasticity simulations. I have run into slowdowns when trying to run larger simulations, so I have tried to profile my code using gperftools and pprof. The only hiccup I had in the installation process was updating and rebuilding libmesh. When running that script, I got the following error:

```
(moose) harrysmith@Harrys-MacBook-Pro moose % scripts/update_and_rebuild_libmesh.sh
ERROR: LIBMESH_DIR=/Users/harrysmith/miniforge/envs/moose/libmesh exists within the conda base /Users/harrysmith/miniforge
Exiting in order to stop you from overwriting your conda environment
```
I proceeded without rebuilding libmesh and I was able to compile and run in oprof mode. I successfully got the profile `.prof` files out, but I am getting this error when I try to use pprof to read the profiles:
```
(moose) harrysmith@Harrys-MacBook-Pro run_4 % pprof run4_0.prof 
Local symbolization failed for libresolv.9.dylib: stat /usr/lib/libresolv.9.dylib: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libvtkIOLegacy-9.2.9.2.5.dylib
Type: cpu
```
I can still see some of the MOOSE framework functions when I use top, but most of the function run time is clumped under `libresolv.9.dylib` , making this profiling tool largely unhelpful:
```
Showing top 10 nodes out of 168
      flat  flat%   sum%        cum   cum%
    33.62s 92.82% 92.82%     36.22s   100%  [libresolv.9.dylib]
     0.30s  0.83% 93.65%      0.30s  0.83%  RankFourTensorTempl::operator*
     0.18s   0.5% 94.15%      0.32s  0.88%  RankTwoTensorTempl::times
```

I have tied setting `PPROF_BINARY_PATH` to the path of the oprof executable file, but that did not help. I found this missing `libresolv.9.dylib` file online, put it in my project directory, and tried setting `PPROF_BINARY_PATH` to the path of that file. I got the same error after trying that. Lastly, I have also tried removing my Conda environment and gperftools and reinstalling them without any luck. 

I am running this on an Apple Silicon Mac, but I have also tried running this on WSL and got the same missing file error. Thank you in advance for your time and help.

**Comment by milljm** [comment]
I don't know much about profiling, but I know we've created a Conda package for it. Just checking for the simple things that I can do before another among my team attempts to answer you;

Are you using our `moose-pprof` Conda package?
```bash
conda activate moose
conda install moose-pprof
```
Next, be sure to build your application using the `oprof` method:
```bash
cd ~/projects/YourApp
METHOD="oprof" make clobberall
METHOD="oprof" make -j 8
...
pprof the_thing
```
> [!TIP]
>There should be no need to rebuild libMesh in preparation to profile your app as we distribute `moose-libmesh` with "profiling" methods already. Our `moose-libmesh` package comes with the following methods: `opt, dbg, devel, oprof`.

**Reply by smithht** [reply]
I just installed the `moose-pprof` package and recompiled using the `oprof` method and I am having the same issue. It looks like I am missing another symbolization file too now:
```
Local symbolization failed for libresolv.9.dylib: stat /usr/lib/libresolv.9.dylib: no such file or directory
Local symbolization failed for libsystem_kernel.dylib: error reading fat Mach-O file /usr/lib/system/libsystem_kernel.dylib: architecture not found in /usr/lib/system/libsystem_kernel.dylib: arm64
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libvtkIOLegacy-9.2.9.2.5.dylib
Type: cpu
```

**Reply by milljm** [reply]
I get the same error:
```pre
❯ pprof run1_1.prof
Local symbolization failed for libresolv.9.dylib: stat /usr/lib/libresolv.9.dylib: no such file or directory
Local symbolization failed for libsystem_kernel.dylib: error reading fat Mach-O file /usr/lib/system/libsystem_kernel.dylib: architecture not found in /usr/lib/system/libsystem_kernel.dylib: arm64
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libvtkIOLegacy-9.2.9.2.5.dylib
Type: cpu
Entering interactive mode (type "help" for commands, "o" for options)
(pprof)  
```
It appears to be a warning of sorts, as I am still able to generate PNG graphs. But I admit, I know nothing about profiling...

I also receive a rather cryptic set of errors upon installing `moose-pprof`. I realize this package is old. Perhaps I simply need to re-build it using the latest compilers `moose-dev` is pinned to.

**Reply by smithht** [reply]
Yeah, I am able to create the PNG graphs and see functions using top, too. The graphs are not very useful though since most of the runtime is lumped under that unsymbolized `libresolv.9.dylib` file. I would like to be able to see the actual functions that are taking up most of the runtime under that cryptic file.

**Reply by milljm** [reply]
Let me build a new version of `moose-pprof`. I am hopeful all those errors during installation are related to `pprof` not functioning correctly. Once complete, I'll post another comment with information on how to add this channel to your Conda list, so you can try to install/use it.

**Reply by milljm** [reply]
I completely forgot about this, I apologize! Working on this now...

https://github.com/idaholab/moose/issues/27751

**Reply by milljm** [reply]
I still get the same issue with an updated version of several packages (compiler, libcxx, etc). Unfortunately this also included the latest version of gperf:

```pre
# Build/Install pprof from google at specified hash
go install github.com/google/pprof@latest
```

I am a bit stumped at the moment.

**Reply by milljm** [reply]
Time to figure this out:

```pre
/mann/xsxp.n']'
   INFO (moose-pprof,pprof/bin/pprof): Needed DSO /opt/MacOSX11.3.sdk/usr/lib/libSystem.B.dylib found in $SYSROOT
   INFO (moose-pprof,pprof/bin/pprof): Needed DSO /opt/MacOSX11.3.sdk/usr/lib/libresolv.9.dylib found in $SYSROOT
   INFO (moose-pprof,pprof/bin/pprof): Needed DSO /opt/MacOSX11.3.sdk/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation found in $SYSROOT
   INFO (moose-pprof,pprof/bin/pprof): Needed DSO /opt/MacOSX11.3.sdk/System/Library/Frameworks/Security.framework/Versions/A/Security found in $SYSROOT
WARNING (moose-pprof): run-exports library package conda-forge/osx-arm64::mpich==4.2.1=h9ee2227_101 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (moose-pprof): run-exports library package conda-forge/osx-arm64::perl==5.32.1=7_h4614cfb_perl5 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (moose-pprof): dso library package conda-forge/osx-arm64::gperftools==2.10=hb7217d7_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```
This _has_ to be important 😏

**Reply by milljm** [reply]
Indeed... this is better:
```bash
❯ pprof run1_1.prof
Local symbolization failed for CommonAuth: stat /System/Library/PrivateFrameworks/CommonAuth.framework/Versions/A/CommonAuth: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libvtkIOLegacy-9.3.9.3.dylib
Type: cpu
Entering interactive mode (type "help" for commands, "o" for options)


(pprof) top 2
Showing nodes accounting for 20ms, 50.00% of 40ms total
Showing top 2 nodes out of 80
      flat  flat%   sum%        cum   cum%
      10ms 25.00% 25.00%       10ms 25.00%  MPIDU_genq_shmem_queue_dequeue
      10ms 25.00% 50.00%       10ms 25.00%  MPIR_Igatherv_allcomm_sched_linear
(pprof)  
```

**Reply by smithht** [reply]
Thank you for your help on this, Jason! It seems like some of the root causes to the symbolization errors are fixed. Are any of the changes you made to `moose-pprof` available for me to try, so I can continue to troubleshoot on my end?

**Reply by milljm** [reply]
I am still trying to shuffle this into our public Conda channel.

But when I ask others in my group why no one has brought up the fact that `moose-pprof` isn't working, they all gave me an answer that it _does_ work.

I am still moving forward with updating pprof, but it's possible something else might be causing your issue. I was told "you have to build libMesh in a special way". But when I read the instructions, that special way is being done automatically these days:


>libMesh automatically adds -fno-omit-frame-pointer to METHOD=oprof builds. However, if you want to do profiling with other methods (e.g. opt, which will not give line number information), both libMesh and MOOSE should be built with CXXFLAGS=-fno-omit-frame-pointer...

ref: https://mooseframework.inl.gov/application_development/profiling.html

I think I'd like to bring @lindsayad in on this, as he might have more insight. I'll certainly post once I can get an updated `moose-pprof` into a location you can grab.

**Reply by lindsayad** [reply, marked as answer]
I have not tried profiling on Mac with gperftools for years (back when they used Intel), but back when I tried the results were always kind of bad compared to Linux. On Mac I used Instruments for profiling

**Reply by smithht** [reply]
Instruments worked well for me. I was able to generate the profile, and used InstrumentsToPprof to convert it to a pprof file. After that, I was able to use all of the pprof tools to view and visualize the profile. Thanks for your help, @milljm and @lindsayad

**Reply by lindsayad** [reply]
that's great to hear @smithht. I had not heard about `InstrumentsToPprof`. Thanks for bringing that to our attention!
